src/v/datalake/record_translator.cc

```diff
@@ -78,19 +97,51 @@ record_translator::build_type(std::optional<resolved_type> val_type) {
 
 ss::future<checked<iceberg::struct_value, record_translator::errc>>
 record_translator::translate_data(
+  model::partition_id pid,
   kafka::offset o,
   iobuf key,
   const std::optional<resolved_type>& val_type,
   iobuf parsable_val,
-  model::timestamp ts) {
+  model::timestamp ts,
+  const chunked_vector<std::pair<std::optional<iobuf>, std::optional<iobuf>>>&
+    headers) {
     auto ret_data = iceberg::struct_value{};
-    ret_data.fields.emplace_back(iceberg::long_value(o));
+    auto system_data = std::make_unique<iceberg::struct_value>();
+    system_data->fields.emplace_back(iceberg::int_value(pid));
+    system_data->fields.emplace_back(iceberg::long_value(o));
     // NOTE: Kafka uses milliseconds, Iceberg uses microseconds.
-    ret_data.fields.emplace_back(iceberg::timestamp_value(ts.value() * 1000));
-    ret_data.fields.emplace_back(iceberg::binary_value{std::move(key)});
+    system_data->fields.emplace_back(
+      iceberg::timestamp_value(ts.value() * 1000));
+
+    auto headers_list = std::make_unique<iceberg::list_value>();
+    for (const auto& [k, v] : headers) {
+        auto header_kv_struct = std::make_unique<iceberg::struct_value>();
+        if (k) {
+            header_kv_struct->fields.emplace_back(
+              iceberg::binary_value(k->copy()));
+        } else {
+            header_kv_struct->fields.emplace_back(std::nullopt);
+        }
```

Comment:
Can't wait for C++23... 

https://en.cppreference.com/w/cpp/utility/optional/and_then

Reply:
Rewrote this to be just a bit less ugly